lib/fast_jsonapi/relationship.rb

```diff
@@ -120,5 +153,74 @@ def run_key_transform(input)
         input.to_sym
       end
     end
+
+    def initialize_static_serializer
+      return if @initialized_static_serializer
+      @initialized_static_serializer = true
+      @static_serializer = compute_static_serializer
+      @static_record_type = compute_static_record_type
+    end
+
+    def compute_static_serializer
+      if serializer.is_a?(Symbol) || serializer.is_a?(String)
+        # a serializer was explicitly specified by name -- determine the serializer class
+        serializer_for_name(serializer)
+
+      elsif serializer.is_a?(Proc)
+        # the serializer is a Proc to be executed per object -- not static
+        nil
+
+      elsif serializer
+        # something else was specified, e.g. a specific serializer class -- return it
+        serializer
+
+      elsif polymorphic
+        # polymorphic without a specific serializer --
+        # the serializer is determined on a record-by-record basis
+        nil
+
+      elsif object_block
+        # an object block is specified without a specific serializer --
+        # assume the objects might be different and infer the serializer by their class
+        nil
+
+      else
+        # no serializer information was provided -- infer it from the relationship name
+        serializer_name = name.to_s
+        serializer_name = serializer_name.singularize if relationship_type.to_sym == :has_many
+        serializer_for_name(serializer_name)
+      end
+    end
+
+    def serializer_for_name(name)
+      @serializers_for_name[name] ||= compute_serializer_for_name(name)
+    end
+
+    def compute_serializer_for_name(name)
+      namespace = owner.name.gsub(/()?\w+Serializer$/, '')
```

Comment:
Would be great to actually use some namespaced fixtures for exactly this purpose.

Reply:
Long story short, I simply moved [compute_serializer_name](https://github.com/fast-jsonapi/fast_jsonapi/blob/c305b699a970ba60e77ba5d7d1f962390bcdc363/lib/fast_jsonapi/object_serializer.rb#L278) from the serializer into the relationship.  And because the relationship is now responsible for computing the serializer name when necessary, and the computation is based on the original serializer's name (for namespacing), we need a reference to the `owner` for this computation.

With this PR, the Relationship's `serializer` property is specifically the `serializer` parameter passed into the `belongs_to` / `has_many` definition.  Before this change, `serializer` was always specified on the Relationship, because if it was not defined, it was computed by the serializer and passed in as `serializer`, and the Relationship could not tell the difference.  This caused confusion for `polymorphic` relationships, because it didn't know whether to use the `serializer` or compute the `polymorphic`.

This way, the Relationship is now solely responsible for determining the serializer for each object in the relationship being serialized (Single Responsibility Principle).

Now, it is all clear.  If `serializer` is specified as a parameter, it will always be used, regardless of `polymorphic`.  If `polymorphic` is specified without `serializer`, it will compute the serializer name.  If neither are defined, it will resolve the _static serializer_ and only compute it once.

As for _how_ the name is computed, I did not change that at all.  I tend to think this PR has enough change as it is.  I think that if we want to address that, we should do so as a separate PR and address it holistically.

Reply:
@christophersansone do you mind if we move the call to the `compute_serializer_for_name` into the `create_relationship`, and pass directly the computed serializer to `options[:serializer] || compute_serializer_for_name(name, owner)`?!

Also ideally, could we make this `compute_serializer_for_name` a public API method pretty please?
The idea is to allow pragmatically an user implement a _renderer_ instead of requiring to provide the serializer class name every time. What do you think?

A real-world rails renderer example could be found here:
https://github.com/stas/jsonapi.rb/blob/3ebf4982243db36d7b5a54ab3c733f76f1df71ba/lib/jsonapi/rails.rb#L100

Reply:
Also here: https://github.com/Netflix/fast_jsonapi/issues/449

Reply:
@stas I can move it, but the relationship must know the difference between when `options[:serializer]` was specified vs. when the serializer name was computed, which it can't do if we are passing in `options[:serializer] || compute...`.  There is enough complexity to the order of operations that need to occur, that in my opinion, it should be the sole responsibility of `Relationship` to determine the serializer for a given relationship / object.  `Relationship` also needs `compute_serializer_for_name` at the object level for `polymorphic: true` relationships.

If your desire here is to simply make `compute_serializer_for_name` a public, static, overridable method on `ObjectSerializer`, I can do that no problem.  Then the relationship can simply call `owner.class.compute_serializer_for_name()` (or whatever) as needed.  Does that address your concern?

Reply:
> I can move it, but the relationship must know the difference between when `options[:serializer]` was specified vs. when the serializer name was computed

I'm not sure I understand, why is this important?

We have these cases here:
 * [ ] a class was provided as an option, no computation is needed
 * [ ] a symbol/string was provided, we need to compute a klass out of it
 * [ ] a proc was provided, no computation is needed, will be called directly
 * [ ] nothing was provided, we use the relationship name and compute it as it would be case nr.2

All seem to be independent from the relationship state, the relationship only needs to know about the class or the proc it needs to use/call, same for the polymorphic option, I just think this will greatly simplify things.

> If so, is `compute_serializer_for_name` the right name for this public method?

Sounds good, we can even make it shorter and just use `serializer_for`. Thank you.

Reply:
Other cases not mentioned above:

* When `polymorphic` is `true` and `serializer` is not specified
* When `polymorphic` is `true` and `serializer` is specified
* When `polymorphic` is a Hash and `serializer` is not specified
* When `polymorphic` is a Hash and `serializer` is specified

I know you feel this is overly complicated and you keep mentioning how a different implementation will "greatly simplify things", but I'm not sure I follow how.  Can you offer some pseudocode to illustrate your point?

Reply:
@christophersansone Here's what I have in mind:
```ruby
Relationship.new(
  ...
  serializer: self.serializer_for(options[:serializer]) unless polymorphic.present?,
  polymorphic: polymorphic.transform_values { |_pv| self.serializer_for(_pv) },
  ...
)

def serializer_for(alias)
  # Will return either a class or a proc or raise some name error
end
```

Assuming we don't need to compute the serializer class if the polymorphic mapping is specified though...

This way we just need  to either use the already computed class or call the proc as part of the relationship.

Reply:
A couple concerns I see with the pseudocode:

1. `Relationship.new` is being called on load, which means that the resulting class of `serializer_for` must either be already loaded or can autoload.  Will that work in all cases?  Usually this type of work is deferred until first use, e.g. ActiveRecord's `belongs_to :user, class_name: 'UserSerializer'`.

2. What happens if both `polymorphic` and `serializer` is specified?  In your logic, `polymorphic` will take precedence.  Granted, I tend to think that both shouldn't be used together, but if they are, I feel like if `options[:serializer]` is specified, it should take precedence in all cases.

3. Ultimately, what does the code look like when a relationship is being serialized / included?  In what other areas of the code do we need to check `if polymorphic.present?` or if `serializer` is a Proc, and if so, execute it?  How often would this be replicated throughout the codebase?

I personally don't think this is any simpler.  I think if you followed your pseudocode to its logical conclusion, you would end up with something remarkably like my implementation. :smile:  My implementation simply encapsulates all of the logic of determining the serializer to use for any given record and exposes a nice clean `Relationship.serializer_for(object, params)`, and it resolves and caches as much as it can up front to ensure that subsequent calls are as fast as possible.

Trying to be helpful here but I can't keep splitting hairs on minor differences of opinion on how a line of code should be written. :smile:  I can move `Relationship.compute_serializer_for_name` to `ObjectSerializer.serializer_for`.  After that, can we just merge it?  From there, you are free to try to simplify it, and you would be able to do a more thorough comparison of the changes you want to make, make sure the tests pass, compare performance, etc.  Does that sound okay?

Reply:
@christophersansone sorry was really just trying to understand the implementation not trying to argue about how it was done. Let's do it as you suggest, and get this merged. We can refine alter on.